Allow alternate openssl and gnutls in all.sh

andresag01 opened the pull request:
* Allow the user to specify alternative openssl, gnutls-cli and
gnutls-serv binaries to execute legacy tests in all.sh e.g. SSLv3 tests.
The tests in all.sh pass cleanly with the following gnutls and openssl versions:
    - OpenSSL: OpenSSL 1.0.2g-fips  1 Mar 2016
    - OpenSSL for legacy tests:  OpenSSL 1.0.1j 15 Oct 2014
    - GnuTLS server: gnutls-serv 3.4.10
    - GnuTLS client: gnutls-cli 3.4.10
    - GnuTLS server for legacy tests: gnutls-serv 3.3.8
    - GnuTLS client for legacy tests: gnutls-cli 3.3.8

* All paths to OpenSSL and GnuTLS binaries can be set through environment variables.
* The `--short` option has been removed from the script.
* all.sh fails if any of the tools is not found.
* Version printing of all the tools is included in output_env.sh.

andresag01 commented:
@yanesca @sbutcher-arm

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -121,6 +158,17 @@ fi
 
+msg "info: $0 configuration"
+echo "MEMORY: $MEMORY"
+echo "SHORT: $SHORT"
```
We've removed the SHORT option, so could you please remove this line.

andresag01 replied:
The patch that removes the `SHORT` option has not been merged for sometime, so for consistency I will leave this here until that happens and if necessary, update either of the PRs as they are merged.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -34,13 +34,26 @@ MEMORY=0
     echo -e "  -m|--memory\t\tAdditional optional memory tests."
     echo -e "  -s|--short\t\tSubset of tests."
     echo -e "  -f|--force\t\tForce the tests to overwrite any modified files."
+    echo -e "     --openssl\t\tOpenSSL to use for most tests."
```
To make it a bit clearer, could you indicate that for each OpenSSL and GNUTLS option, a parameter has to be supplied. e.g..
```
echo -e "     --openssl=<OpenSSL path>\t\tPath to OpenSSL executable to use for most tests."
```

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -121,6 +158,17 @@ fi
 #
```
I suggest if the user has specified a specific OpenSSL or GNUTLS version, we should check we can find them in the path (`hash`?) else report an error. Otherwise if we don't, what error will be provided?

andresag01 replied:
Ok, I will do this.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -34,13 +34,26 @@ MEMORY=0
 SHORT=0
 FORCE=0
 
+OPENSSL="openssl"
```
I think we should use existing environment variables to override these options, so we don't have to enter the full path for OpenSSL and GNUTLS at the command line, everytime we execute, so can we make the assignment to <tool>_LEGACY dependent on it not already being defined?

sbutcher-arm replied:
Also, by adding environment variables for OpenSSL and GNUTLS, please also add outputting the version of those legacy versions to the `output_env.sh` script, if they are defined.

andresag01 replied:
Ok, you mean leave the command line option there as it is, but also make it possible to define all these values through environment variables?

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -34,13 +34,26 @@ MEMORY=0
 FORCE=0
 
+OPENSSL="openssl"
+OPENSSL_LEGACY="$OPENSSL"
```
I'm not sure there's much point defaulting to existing OpenSSL as all modern versions (which you should have installed) won't support the older features we're trying to test. It's probably best to fail - because no legacy versions have been specified - and then provide an override option of '--no-legacy-tests' to skip these tests.

andresag01 replied:
I disagree with this, when I introduced the possibility to change both options, I was thinking that someone just has installed an older version of OpenSSL and GnuTLS and would only like to test using those (which would be the default). In that case you would still like to run all tests but your `legacy` and newer version are effectively the same. Also, notice that adding the `--no-legacy-tests` is not straight forward because the tests in `compat.sh` and `ssl-opt.sh` do not have any version requirements and legacy and other tests are mixed in the same run of the scripts.

sbutcher-arm commented:
Also - as part of this pull request, could you please remove lines 8 - 10 of `tests/compat.sh`, which specify incorrectly minimum versions of GNUTLS and OpenSSL.

andresag01 commented:
@sbutcher-arm: I believe I have addressed most of your comments. Notice that because some of the feedback was regarding the SHORT option and fail if tool is not found, I decided to merge PR https://github.com/ARMmbed/mbedtls/pull/580 here as well.

sbutcher-arm commented:
Reviewed, tested, fixed and merged manually.
